File: coreui/src/main/java/com/mbg/module/ui/activity/BaseFragmentActivity.java
package com.mbg.module.ui.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;

import com.mbg.module.ui.fragment.BaseFragment;

import java.util.ArrayDeque;


public abstract class BaseFragmentActivity extends BaseActivity {

    protected static final String ARG_FRAGMENT_CLASS_NAME="arg_fragment_class_name";
    protected static final String ARG_FRAGMENT_ARGS="arg_fragment_args";
    private ArrayDeque<BaseFragment> mFragments = new ArrayDeque<>();


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Intent intent=getIntent();

        Class<? extends BaseFragment> fragmentClass=null;
        Bundle bundle=null;
        if (intent!=null){

            String fragmentName=intent.getStringExtra(ARG_FRAGMENT_CLASS_NAME);
            if (fragmentName!=null){
                try {
                    fragmentClass=(Class<? extends BaseFragment>)getClassLoader().loadClass(fragmentName);
                }catch (Exception e){
                    e.printStackTrace();
                    finish();
                }
            }

            bundle=intent.getBundleExtra(ARG_FRAGMENT_ARGS);
        }

        if (fragmentClass!=null) {
            showContent(fragmentClass, bundle);
        }else {
            finish();
        }
    }

    public void showContent(Class<? extends BaseFragment> target) {
        showContent(target, null);
    }

    public void showContent(Class<? extends BaseFragment> target, Bundle bundle) {
        try {
            BaseFragment fragment = target.newInstance();
            if (bundle != null) {
                fragment.setArguments(bundle);
            }
            FragmentManager fm = getSupportFragmentManager();
            FragmentTransaction fragmentTransaction = fm.beginTransaction();
            fragmentTransaction.add(android.R.id.content, fragment);
            mFragments.push(fragment);
            fragmentTransaction.addToBackStack("");
            fragmentTransaction.commit();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
           e.printStackTrace();
        }
    }

    @Override
    public void onBackPressed() {
        if (!mFragments.isEmpty()) {
            BaseFragment fragment = mFragments.getFirst();
            if (!fragment.onBackPressed()) {
                mFragments.removeFirst();
                super.onBackPressed();
                if (mFragments.isEmpty()) {
                    finish();
                }
            }
        } else {
            super.onBackPressed();
        }
    }

    public void doBack(BaseFragment fragment) {
        if (mFragments.contains(fragment)) {
            mFragments.remove(fragment);
            FragmentManager fm = getSupportFragmentManager();
            fm.popBackStack();
            if (mFragments.isEmpty()) {
                finish();
            }
        }
    }
}

File: coreui/src/main/java/com/mbg/module/ui/kotlin/activity/BaseFragmentActivity.kt
package com.mbg.module.ui.kotlin.activity


import android.os.Bundle
import com.mbg.module.ui.kotlin.fragment.BaseFragment
import java.util.*

abstract class BaseFragmentActivity : BaseActivity() {
    private val mFragments = ArrayDeque<BaseFragment>()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val intent = intent
        var fragmentClass: Class<out BaseFragment>? = null
        var bundle: Bundle? = null
        if (intent != null) {
            val fragmentName = intent.getStringExtra(ARG_FRAGMENT_CLASS_NAME)
            if (fragmentName != null) {
                try {
                    fragmentClass = classLoader.loadClass(fragmentName) as Class<out BaseFragment>
                } catch (e: Exception) {
                    e.printStackTrace()
                    finish()
                }
            }
            bundle = intent.getBundleExtra(ARG_FRAGMENT_ARGS)
        }
        if (fragmentClass != null) {
            showContent(fragmentClass, bundle)
        } else {
            finish()
        }
    }

    @JvmOverloads
    fun showContent(target: Class<out BaseFragment>, bundle: Bundle? = null) {
        try {
            val fragment = target.newInstance()
            if (bundle != null) {
                fragment.arguments = bundle
            }
            val fm = supportFragmentManager
            val fragmentTransaction = fm.beginTransaction()
            fragmentTransaction.add(android.R.id.content, fragment)
            mFragments.push(fragment)
            fragmentTransaction.addToBackStack("")
            fragmentTransaction.commit()
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
    }

    override fun onBackPressed() {
        if (!mFragments.isEmpty()) {
            val fragment = mFragments.first
            if (!fragment.onBackPressed()) {
                mFragments.removeFirst()
                super.onBackPressed()
                if (mFragments.isEmpty()) {
                    finish()
                }
            }
        } else {
            super.onBackPressed()
        }
    }

    fun doBack(fragment: BaseFragment) {
        if (mFragments.contains(fragment)) {
            mFragments.remove(fragment)
            val fm = supportFragmentManager
            fm.popBackStack()
            if (mFragments.isEmpty()) {
                finish()
            }
        }
    }

    companion object {
         const val ARG_FRAGMENT_CLASS_NAME = "arg_fragment_class_name"
         const val ARG_FRAGMENT_ARGS = "arg_fragment_args"
    }
}